File: libgomp/testsuite/libgomp.c-c++-common/lastprivate-conditional-3.c
/* { dg-do run } */
/* { dg-require-effective-target tls_runtime } */
/* { dg-additional-options "-std=gnu99" {target c } } */

#include <omp.h>
#include <stdlib.h>

int r, s, u, v, t;
int *x;

void
foo (int *a)
{
  int i;
  long long j;
  #pragma omp for lastprivate (conditional: u, x) ordered
  for (i = 15; i < 64; i++)
    {
      #pragma omp critical
      {
	if ((a[i] % 5) == 3)
	  u = i;
      }
      #pragma omp ordered
      {
	if ((a[i] % 7) == 2)
	  x = &a[i];
      }
    }
  #pragma omp for lastprivate (conditional: v) reduction (+:r, s) schedule (nonmonotonic: static) reduction (task, +: t)
  for (i = -3; i < 119; i += 2)
    {
      ++s;
      #pragma omp taskgroup
      {
	#pragma omp task in_reduction (+: t)
	  ++t;
	if ((a[i + 4] % 11) == 9)
	  v = i;
	else
	  ++r;
      }
    }
}

int
main ()
{
  int a[128], i;
  for (i = 0; i < 128; i++)
    a[i] = i;
  #pragma omp parallel
  foo (a);
  if (u != 63 || v != 115 || x != &a[58] || r != 55 || s != 61 || t != 61)
    abort ();
  return 0;
}

File: gcc/testsuite/gfortran.dg/gomp/lastprivate-conditional-3.f90
subroutine foo
  integer i, j, k
  !$omp parallel
    !$omp do lastprivate (conditional: i)	! { dg-warning "conditional 'lastprivate' on loop iterator 'i' ignored" }
    do i = 1, 32
    end do
    !$omp do collapse (3) lastprivate (conditional: i)	! { dg-warning "conditional 'lastprivate' on loop iterator 'i' ignored" }
    do i = 1, 32
      do j = 1, 32
        do k = 1, 32
        end do
      end do
    end do
    !$omp do collapse (3) lastprivate (conditional: j)	! { dg-warning "conditional 'lastprivate' on loop iterator 'j' ignored" }
    do i = 1, 32
      do j = 1, 32
        do k = 1, 32
        end do
      end do
    end do
    !$omp do collapse (3) lastprivate (conditional: k)	! { dg-warning "conditional 'lastprivate' on loop iterator 'k' ignored" }
    do i = 1, 32
      do j = 1, 32
        do k = 1, 32
        end do
      end do
    end do
  !$omp end parallel

  !$omp parallel do lastprivate (conditional: i)  ! { dg-warning "conditional 'lastprivate' on loop iterator 'i' ignored" }
  do i = 1, 32
  end do
  !$omp end parallel do

  !$omp parallel do collapse (3) lastprivate (conditional: i)  ! { dg-warning "conditional 'lastprivate' on loop iterator 'i' ignored" }
  do i = 1, 32
    do j = 1, 32
      do k = 1, 32
      end do
    end do
  end do
  !$omp end parallel do

  !$omp parallel do collapse (3) lastprivate (conditional: j)  ! { dg-warning "conditional 'lastprivate' on loop iterator 'j' ignored" }
  do i = 1, 32
    do j = 1, 32
      do k = 1, 32
      end do
    end do
  end do
  !$omp end parallel do

  !$omp parallel do collapse (3) lastprivate (conditional: k)  ! { dg-warning "conditional 'lastprivate' on loop iterator 'k' ignored" }
  do i = 1, 32
    do j = 1, 32
      do k = 1, 32
      end do
    end do
  end do
  !$omp end parallel do
end subroutine
